### PR TITLE
Wen restart: generate full snapshot if no full snapshot is available.

### PR DESCRIPTION
**Pull request by wen-coding**
When the validator isn't configured to generate snapshot or the full snapshots aren't available for any reason, generate full snapshot instead of incremental snapshot.

### CODE REVIEW

File: wen-restart/src/wen_restart.rs [hunk 1]
```diff
@@ -529,34 +520,53 @@ pub(crate) fn generate_snapshot(
...
+        )?
+        .path()
+        .display()
+        .to_string()
+    };
     let new_shred_version =
         compute_shred_version(&genesis_config_hash, Some(&new_root_bank.hard_forks()));
-    let new_snapshot_path = archive_info.path().display().to_string();
-    info!("wen_restart incremental snapshot generated on {new_snapshot_path} base slot {full_snapshot_slot}");
+    info!("wen_restart incremental snapshot generated on {new_snapshot_path} base slot {full_snapshot_slot:?}");
```
Comment (brooksprumo) [inline wen-restart/src/wen_restart.rs]: nit: Looks like this log assumes we're generating an incremental snapshot, even though now we might generate a full snapshot.

**Review by brooksprumo — COMMENTED** [review]
This looks good for the case when we don't have a full snapshot. However, we do also need to handle the separate situation when validators disable generating snapshots.

In this case the validator will have a full snapshot on disk—that they loaded from—but we cannot generate a new incremental snapshot based on it. This is because the state of zero lamports will break invariants required by incremental snapshots. Instead, if the validator has snapshot generation turned off, we want to generate a new full snapshot.

File: wen-restart/src/wen_restart.rs (hunk 1 above)
Comment (wen-coding) [reply wen-restart/src/wen_restart.rs]: Done.

**Comment by wen-coding** [comment]
> This looks good for the case when we don't have a full snapshot. However, we do also need to handle the separate situation when validators disable generating snapshots.
> 
> In this case the validator will have a full snapshot on disk—that they loaded from—but we cannot generate a new incremental snapshot based on it. This is because the state of zero lamports will break invariants required by incremental snapshots. Instead, if the validator has snapshot generation turned off, we want to generate a new full snapshot.

Ah, I misunderstood you, changed.

File: wen-restart/src/wen_restart.rs [hunk 2]
```diff
@@ -3343,6 +3369,31 @@ mod tests {
...
             WenRestartError::BlockNotFound(empty_slot),
         );
+        // Now turn off snapshot generation, we should generate a full snapshot.
+        let snapshot_config = SnapshotConfig {
+            bank_snapshots_dir: bank_snapshots_dir.as_ref().to_path_buf(),
+            full_snapshot_archives_dir: full_snapshot_archives_dir.as_ref().to_path_buf(),
+            incremental_snapshot_archives_dir: incremental_snapshot_archives_dir
+                .as_ref()
+                .to_path_buf(),
+            usage: SnapshotUsage::LoadOnly,
```
Comment (brooksprumo) [inline wen-restart/src/wen_restart.rs:3379]: Nice, thanks for the added test here!

**Review by brooksprumo — APPROVED** [review]
:shipit: 

Code looks good. A subsequent PR may be useful to inspect-and-update the various comments that things like "we always generate an incremental snapshots". Not necessary to fix in this PR though.

**Pull request merged**